File: prototype/rails/src/test/domain/orders/order_payment_test.rb
require "test_helper"

class OrderPaymentTest < ActiveSupport::TestCase
  Payment = Domain::Orders::OrderPayment
  Status = Domain::Orders::OrderStatus

  test "an order awaiting payment takes a card" do
    assert Payment.awaits_card?(Status::AWAITING_PAYMENT)
  end

  test "a declined order takes another card" do
    assert Payment.awaits_card?(Status::PAYMENT_FAILED)
  end

  test "an unverified order takes no card yet" do
    refute Payment.awaits_card?(Status::PENDING_VERIFICATION)
  end

  test "a paid order takes no card" do
    refute Payment.awaits_card?(Status::PAID)
  end

  test "an unverified order is unpaid" do
    assert Payment.unpaid?(Status::PENDING_VERIFICATION)
  end

  test "a declined order is unpaid" do
    assert Payment.unpaid?(Status::PAYMENT_FAILED)
  end

  test "a shipped order is not unpaid" do
    refute Payment.unpaid?(Status::SHIPPED)
  end

  test "a verified purchaser pays an order awaiting payment" do
    assert Payment.payable?(Status::AWAITING_PAYMENT, true)
  end

  test "an unverified purchaser pays nothing" do
    refute Payment.payable?(Status::AWAITING_PAYMENT, false)
  end

  test "a verified purchaser cannot pay a delivered order" do
    refute Payment.payable?(Status::DELIVERED, true)
  end
end

File: prototype/node/src/app/core/orders/order-payment.test.ts
import { test } from 'node:test'
import assert from 'node:assert/strict'
import { awaitsCard, isUnpaid, isPayable } from './order-payment.ts'

test('an order awaiting payment takes a card', () => {
  assert.equal(awaitsCard('awaiting_payment'), true)
})

test('a declined order takes another card', () => {
  assert.equal(awaitsCard('payment_failed'), true)
})

test('an unverified order takes no card yet', () => {
  assert.equal(awaitsCard('pending_verification'), false)
})

test('a paid order takes no card', () => {
  assert.equal(awaitsCard('paid'), false)
})

test('an unverified order is unpaid', () => {
  assert.equal(isUnpaid('pending_verification'), true)
})

test('a declined order is unpaid', () => {
  assert.equal(isUnpaid('payment_failed'), true)
})

test('a shipped order is not unpaid', () => {
  assert.equal(isUnpaid('shipped'), false)
})

test('a verified purchaser pays an order awaiting payment', () => {
  assert.equal(isPayable('awaiting_payment', true), true)
})

test('an unverified purchaser pays nothing', () => {
  assert.equal(isPayable('awaiting_payment', false), false)
})

test('a verified purchaser cannot pay a delivered order', () => {
  assert.equal(isPayable('delivered', true), false)
})
